instrumentation/grpc/lib/opentelemetry/instrumentation/grpc/patches/active_call.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module GRPC
+      module Patches
+        # Patches to prepend to ::GRPC::ActiveCall.
+        module ActiveCall
+          def initialize(call, marshal, unmarshal, deadline, started: true, metadata_received: false, metadata_to_send: nil)
+            super
+            @span = OpenTelemetry::Trace.current_span
+          end
+
+          # We need a setter because while we can easily capture a correct trace during initialization
+          # on the client side, it's a little harder in the accept loop on the server. So, we provide
+          # a way for server-side ActiveCalls to set the span after the ActiveCall is created.
+          def span=(span)
+            @span = span
```

Comment:
I am unfamiliar with the execution model here. Does this require a threadsafe writer?

Reply:
I don't _think_ so. I will admit that fundamentally I am not 100% sure. This ActiveCall setter is only called in two places:
- when the ActiveCall is initialized (which sets the "correct" span for client side things using the ActiveCall class)
- when we start to handle a call on the server-side - just after it is dispatched for processing, but technically _after_ the ActiveCall was already created.

The main reason I added it was because on the server-side, I couldn't find a great way to cleanly patch the accept loop that handles and schedules all of the server methods without re-implementing a lot of that code myself. That felt like a great way to introduce bugs to me 😆.

The "accept loop" (so to speak) is at [`RpcServer#loop_handle_server_calls`](https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/rpc_server.rb#L448) - and the bit I did choose to patch was [`RpcDesc#run_server_method`](https://github.com/grpc/grpc/blob/master/src/ruby/lib/grpc/generic/rpc_desc.rb#L117), which is called from within `RpcServer#loop_handle_server_calls`. If you take a peek at the handling code, you can probably see why I chose to do it this way. However, a fresh set of eyes would be appreciated sometime, because I may well have overlooked the easy way to do this 🤣 

Anyways - back to the thread safety question - the reason I think this is okay is because we only ever change this variable in one place, and that specifically is when we dispatch the server-side ActiveCall to a thread for processing. A different ActiveCall is created for each request coming in, and my read of the work-processing system within the GRPC code is that requests are explicitly scheduled _with_ their context to a single thread, rather than multiple threads potentially pulling the same work item or something. I don't immediately see a way in which there could be concurrent access to this setter given the scope and the overall way that the GRPC library is doing things.

That said ... the GRPC library is rather confusing, so I very much could be wrong there. I'm going to say I'm about 95% certain we are fine here, in my opinion.